lib/src/source/hosted.dart

```diff
@@ -167,6 +168,57 @@ class HostedSource extends CachedSource {
     }
   }();
 
+  /// Whether extra metadata headers should be sent for HTTP requests to a given
+  /// [url].
+  static bool _shouldSendRequestMetadata(Uri url) {
+    if (runningFromTest && Platform.environment.containsKey('PUB_HOSTED_URL')) {
+      if (url.origin != Platform.environment['PUB_HOSTED_URL']) {
+        return false;
+      }
+    } else {
+      if (!HostedSource.isPubDevUrl(url.toString())) return false;
+    }
+
+    if (Platform.environment.containsKey('CI') &&
+        Platform.environment['CI'] != 'false') {
+      return false;
+    }
+
+    return true;
+  }
+
+  /// Adds request metadata headers to a [Map] of HTTP headers. The headers
+  /// contain information about the Pub tool's environment and the currently
+  /// running command.
+  static void attachRequestMetadata(Map<String, String> headers) {
+    headers['X-Pub-OS'] = Platform.operatingSystem;
+    headers['X-Pub-Command'] = PubCommand.command;
+    headers['X-Pub-Session-ID'] = sessionId;
+
+    var environment = Platform.environment['PUB_ENVIRONMENT'];
+    if (environment != null) {
+      headers['X-Pub-Environment'] = environment;
+    }
+
+    var type = Zone.current[#_dependencyType];
```

Comment:
We can't use this symbol here, I at-least I don't think we can. Since it's a private symbol, it should be distinct from the symbol defined in `http.dart`:
https://github.com/dart-lang/pub/blob/86536fd07dc62705a2b1b3eb46e1208b3a361f3a/lib/src/http.dart#L138

At-least that's my guess.

------------------------------------------------------

May I suggest that we make:
```dart
extension AttachHeaders on http.Request {
  void attachMetadataHeaders() {
      ...
  }
}
```

in `http.dart`.

We'll have to refactor the places that use `Client.get` and `Client.read` into creating `Request` objects  and passing them to `Client.send`.
But I don't think that matters much.

Reply:
Makes me wonder if we have any tests covering the meta-data sending?

It would probably require changing `isPubDevUrl` to something like:

```dart
  static bool isPubDevUrl(String url) {
    final origin = Uri.parse(url).origin;

    return origin == pubDevUrl ||
        origin == pubDartlangUrl ||
        (runningFromTest &&
            origin == io.Platform.environment['_PUB_TEST_DEFAULT_HOSTED_URL']);
  }
```

Reply:
I've changed the headers method to an extension in http.dart. 👍

Re: the metadata test, I'd like to do that test but need some clarification on the desired behavior for the metadata URL-checking logic. I've sent you a message on Chat.

Reply:
How about we do tests for metadata sending a follow up PR (it's not tests we currently have). And it's probably better to keep this PR small.